src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParser.cs

```diff
@@ -147,14 +147,31 @@ private TypeNameParser(ReadOnlySpan<char> name, bool throwOnError, TypeNameParse
             {
                 _inputString = capturedBeforeProcessing;
             }
+            else
+            {
+                // Every constructed generic type needs the generic type definition.
+                if (!TryDive(_parseOptions, ref recursiveDepth))
+                {
+                    return null;
+                }
+                // If that definition represents a nested type, this needs to be taken into account.
+                if (nestedNameLengths is not null)
+                {
+                    recursiveDepth += nestedNameLengths.Count;
```

Comment:
Should this be done earlier as soon as we discover that the nesting is too deep?

Also, I am having hard time understanding why the `TryDive` at https://github.com/dotnet/runtime/pull/106334/files#diff-905542846dc45d4936395b08bdac3b71d1f5e03529a188298a93f20d080c19c3R252 is not sufficient to handle this case.

Reply:
> Should this be done earlier as soon as we discover that the nesting is too deep?

Ideally yes, but at that moment of time we don't know whether the name that is being parsed represents a generic type. 

> Also, I am having hard time understanding why the TryDive at is not sufficient to handle this case.

This `TryDive` increases the node count once, but if it turns out that it's a generic type, we need to do that again (because the generic type definition is also going to be  a nested type and will require the same amount of type name instances.

I've added following comment to the code, please let me know if this is enough:

```cs
// If that generic type is a nested type, this needs to be taken into account. Example:
// "Namespace.Declaring+NestedGeneric`1[GenericArg]" requires 5 TypeName instances:
// - constructed generic type: Namespace.Declaring+NestedGeneric`2[GenericArg] (+1, handled by ParseNextTypeName)
//   - declaring type: Namespace.Declaring (+1, handled by TryGetTypeNameInfo)
//   - generic type definition: Namespace.Declaring+NestedGeneric`1 (+1, handled by the TryDive above)
//      - declaring type: Namespace.Declaring (+1, handled by the if below)
//   - generic arguments:
//     - simple: GenericArg (+1, handled by ParseNextTypeName)
```